test/integration_test/basic_test.go

```diff
@@ -6,186 +6,206 @@ import (
-	require.NoError(t, err)
-}
+var (
+	pxVer2_9, _ = version.NewVersion("2.9")
+)
 
-func testNodeAffinityLabels(t *testing.T) {
-	var err error
-	labelKey := "skip/px"
-	labelValue := "true"
-
-	// Get versions from URL
-	logrus.Infof("Get component images from versions URL")
-	imageListMap, err := testutil.GetImagesFromVersionURL(pxSpecGenURL)
-	require.NoError(t, err)
-
-	// Construct Portworx StorageCluster object
-	cluster, err := constructStorageCluster(pxSpecGenURL, imageListMap)
-	require.NoError(t, err)
-
-	cluster.Name = "node-affinity-labels"
-
-	// Set Node Affinity
-	cluster.Spec.Placement = &op_corev1.PlacementSpec{
-		NodeAffinity: &v1.NodeAffinity{
-			RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
-				NodeSelectorTerms: []v1.NodeSelectorTerm{
-					{
-						MatchExpressions: []v1.NodeSelectorRequirement{
-							{
-								Key:      labelKey,
-								Operator: v1.NodeSelectorOpNotIn,
-								Values:   []string{labelValue},
+var testStorageClusterBasicCases = []TestCase{
+	{
+		TestName:        "InstallWithAllDefaults",
+		TestrailCaseIDs: []string{},
+		TestStorageCluster: &op_corev1.StorageCluster{
+			ObjectMeta: meta.ObjectMeta{Name: "simple-install"},
+		},
+		ShouldSkip: false,
+		TestFunc:   BasicInstall,
+	},
+	{
+		TestName:        "NodeAffinityLabels",
+		TestrailCaseIDs: []string{},
+		TestStorageCluster: &op_corev1.StorageCluster{
+			ObjectMeta: meta.ObjectMeta{Name: "node-affinity-labels"},
+			Spec: op_corev1.StorageClusterSpec{
+				Placement: &op_corev1.PlacementSpec{
+					NodeAffinity: &v1.NodeAffinity{
+						RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+							NodeSelectorTerms: []v1.NodeSelectorTerm{
+								{
+									MatchExpressions: []v1.NodeSelectorRequirement{
+										{
+											Key:      labelKeySkipPX,
+											Operator: v1.NodeSelectorOpNotIn,
+											Values:   []string{labelValueTrue},
+										},
+									},
+								},
 							},
 						},
 					},
 				},
 			},
 		},
-	}
+		ShouldSkip: false,
+		TestFunc:   BasicInstallWithNodeAffinity,
+	},
+	{
+		TestName:        "Upgrade",
+		TestrailCaseIDs: []string{},
+		TestStorageCluster: &op_corev1.StorageCluster{
+			ObjectMeta: meta.ObjectMeta{Name: "upgrade-test"},
+		},
+		ShouldSkip: false,
+		TestFunc:   BasicUpgrade,
+	},
+}
 
-	// Get K8S nodes
-	nodeList, err := coreops.Instance().GetNodes()
-	require.NoError(t, err)
+func TestStorageClusterBasic(t *testing.T) {
+	for _, testCase := range testStorageClusterBasicCases {
+		testCase.RunTest(t)
+	}
+}
 
-	// Set Node Affinity label one of the K8S nodes
-	var nodeNameWithLabel string
-	for _, node := range nodeList.Items {
-		if coreops.Instance().IsNodeMaster(node) {
-			continue // Skip master node, we don't need to label it
+func BasicInstall(tc *TestCase) func(*testing.T) {
+	return func(t *testing.T) {
+		if tc.ShouldSkip {
```

Comment:
Can ShouldSkip be a function returning bool instead of a bool value itself? That way every test case can have its own skip function or share a common one applicable to different environments.
Also, should we just skip before even calling this test func.. before or inside RunTest method?

Reply:
we could make ShouldSkip a function when defining the test case, but seems there's some issue as mentioned in Upgrade function like the global param was not initialized. but think your suggestion would probably fix that issue so I'll try.
also the t.Skip function only takes args not test name, think about our previous test functions, the skip should be called inside, and the test func is called by t.Run(), so I don't know if this could be called outside of t.Run()

Reply:
We don't need to call t.Skip(), right? We can just not call the testfunc if the shouldSkip function returns true.

Reply:
yeah we could do that, only thing is the skipped tests will not printed in the console at the end of test run

Reply:
Ah, I see. Then we could call the t.skip inside the test as you have done now.

Reply:
yes we could, that way the code is simpler but we don't have a consistent golang test report, but eventually we need to have our own report to make it easier, e.g. print skipped/passed/failed tests and their testrails IDs. I'm fine with change it now or later when we reach there @piyush-nimbalkar what do you think?

Reply:
I am fine with either approach. Would have liked to keep the tests cleaner with only test logic, and have the TestCase struct and loop take care of other things.
If you think the reporting is better this way for now keep it inside the test function or if you are going to keep it outside, just add a log line saying we are skipping xx test.